### ISSUE TITLE
Nested events and boomeranging states

### ISSUE THREAD

**Issue by steveluscher** [pluginaweek/state_machine#201]
Start with ActiveRecord. Let's say that you fire an event, and that during that event you fire another event. Let's say you're dealing with an Order model that has a checkout event, during which payment is processed:

```
class Order < ActiveRecord::Base
  state_machine :state, initial: :incomplete do
    state :processing_payment
    state :paid

    after_transition :incomplete => :processing_payment, do: :process_payment

    event :checkout do
      transition :incomplete => :processing_payment
    end
    event :mark_as_paid do
      transition :processing_payment => :paid
    end
    event :mark_as_having_failed_payment do
      transition :processing_payment => :incomplete
    end
  end
end
```

Let's say that, in `process_payment`, you discover the buyer's card has been declined, and you want to call `mark_as_having_failed_payment`.

```
def process_payment
  …
  mark_as_having_failed_payment! if payment_failed
end
```

I would expect the record to now have the state `incomplete`, but it has the state `processing_payment` instead. Here's the log from the transition between `:incomplete` and `:processing_payment`:

```
(0.4ms)  UPDATE `orders` SET `state` = 'processing_payment', `updated_at` = '2012-07-30 00:16:20' WHERE `orders`.`id` = 123
```

And here's the log generated as a result of calling `mark_as_having_failed_payment!`:

```
(1.1ms)  UPDATE `orders` SET `updated_at` = '2012-07-30 00:16:20' WHERE `orders`.`id` = 123
```

Why didn't it set the state? It's as though the machine _thinks_ that since it started in the `:incomplete` state, there's nothing to do, even though the previous event has already set the state to `:processing_payment` _in the same transaction_.

Note that this is not a problem if you're going to a different state altogether (like `:paid` for example). It's only a problem when you go from state A => B => A.

**Comment by the8472** [comment]
While this might a bug I don't think it's The Right Way(tm) to do it.

You could just use a `before_transition` handler and return false to rollback the transition.
Another way would be to determine the target state beforehand by using an event with multiple, conditional transitions

**Comment by obrie** [comment]
Thanks for reporting this!  I've revamped the way state_machine integrates with ActiveRecord -- this has subsequently fixed this issue.  It'll get included in the upcoming 1.2.0 release.

**Comment by steveluscher** [comment]
Awesome. Cheers!
